### PR TITLE
[DO NOT MERGE] Adding routing properties

### PR DESCRIPTION
**Pull request by rkmanda**
This checklist is used to make sure that common issues in a pull request are addressed. This will expedite the process of getting your pull request merged and avoid extra work on your part to fix issues discovered during the review process. 

### PR information
- [ ] The title of the PR is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).
- [ ] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [ ] If applicable, the PR references the bug/issue that it fixes.
- [ ] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).

### Quality of Swagger
- [ ] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [ ] My spec meets the review criteria:
  - [ ] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [ ] Validation errors from the [Linter extension for VS Code](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/linter.md) have all been fixed for this spec. (**Note:** for large, previously checked in specs, there will likely be many errors shown. Please contact our team so we can set a timeframe for fixing these errors if your PR is not going to address them).
  - [ ] The spec follows the patterns described in the [Swagger good patterns](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-good-patterns.md) document unless the service API makes this impossible.

### CODE REVIEW

File: arm-iothub/2016-02-03/swagger/iothub.json
Comment (amarzavery) [inline arm-iothub/2016-02-03/swagger/iothub.json]: Is adding a new feature to an existing api-version, allowed? Wouldn't this be a breaking change in the sdk?

Comment (amarzavery) [inline arm-iothub/2016-02-03/swagger/iothub.json]: Looks like this is an optional property. So if someone make a request to get an IotHub, 
 - how will this property look like? 
   - Will it not be present at all?
   - Will it be present with a null value ?
   - Will it be present with some of the properties (readOnly), set ?
   - Or will it be an empty object?

File: arm-iothub/2016-02-03/swagger/iothub.json
Comment (amarzavery) [inline arm-iothub/2016-02-03/swagger/iothub.json]: anything required here? Based on the description of the fallbackRoute, looks like it needs to be provided, as that will govern the behavior if none of the routing rule conditions are met from the rules provided in the routes array?

File: arm-iothub/2016-02-03/swagger/iothub.json
Comment (amarzavery) [inline arm-iothub/2016-02-03/swagger/iothub.json]: anything required here?

File: arm-iothub/2016-02-03/swagger/iothub.json
Comment (amarzavery) [inline arm-iothub/2016-02-03/swagger/iothub.json]: > "The identifier of the endpoint. This will be set by the server. Modifying this value causes create or update IoT hub operations to fail. Pass id as null when specifying a new endpoint. Pass the same id that you get for an endpoint from the service when specifying an existing endpoint during update operations."

The above description is contradicting the `"readOnly": true` set for that property. If the property is readOnly then it will never be sent on the wire (never serialized). It will only be server populated.

Moreover, the first sentence in the description contradicts with the second sentence. 

Plus it is an extremely bad idea, to model the behavior based on null. How will the sdk know, that the property was null because of default object initialization in the constructor or the customer explicitly set it. 

How was this kind of design ever approved by the API review board?

/cc @ravbhatnagar - We need some rules by the ARM team with regards to setting null values for properties.

File: arm-iothub/2016-02-03/swagger/iothub.json
Comment (amarzavery) [inline arm-iothub/2016-02-03/swagger/iothub.json]: same as above

**Review by amarzavery — CHANGES_REQUESTED** [review]

File: arm-iothub/2016-02-03/swagger/iothub.json
Comment (rkmanda) [inline arm-iothub/2016-02-03/swagger/iothub.json]: It will be an empty object

File: arm-iothub/2016-02-03/swagger/iothub.json
Comment (rkmanda) [inline arm-iothub/2016-02-03/swagger/iothub.json]: nope its optional. I will add more description on whats expected when fallbackroute is not specified.

File: arm-iothub/2016-02-03/swagger/iothub.json
Comment (rkmanda) [inline arm-iothub/2016-02-03/swagger/iothub.json]: nope

File: arm-iothub/2016-02-03/swagger/iothub.json
Comment (rkmanda) [inline arm-iothub/2016-02-03/swagger/iothub.json]: Will remove the readOnly property. I was not aware that it would actually remove the property being passed in. Was considering it as an indication to the user to not modify the value. Also, we are aware and debated at length regarding the weirdness with this behavior. However, there are reasons for doing it the way it is being done. We can have a 1:1 conversation if you care to understand. Btw, who is the "API review board" ?

File: arm-iothub/2016-02-03/swagger/iothub.json
Comment (rkmanda) [inline arm-iothub/2016-02-03/swagger/iothub.json]: same as above

File: arm-iothub/2016-02-03/swagger/iothub.json
Comment (rkmanda) [inline arm-iothub/2016-02-03/swagger/iothub.json]: can u remove the "breaking change" tag? or explain what this breaks?

File: arm-iothub/2016-02-03/swagger/iothub.json
Comment (rkmanda) [inline arm-iothub/2016-02-03/swagger/iothub.json]: does this description make more sense to you (minus the readonly property) ?

The identifier of the endpoint. An identifier will be assigned by the server, and once assigned the identifier should never be modified. Modifying this value causes update IoT hub operations to fail. Pass id as null or a whitespace when specifying a new endpoint. Pass the same id that you get for an endpoint from the service when specifying an existing endpoint during update operations.

File: arm-iothub/2016-02-03/swagger/iothub.json
Comment (johanste) [inline arm-iothub/2016-02-03/swagger/iothub.json]: Please avoid using NULL as semantically meaningful values. It leads to *very* bad experiences when using the API - particularly for strongly typed languages.

File: arm-iothub/2016-02-03/swagger/iothub.json
Comment (johanste) [inline arm-iothub/2016-02-03/swagger/iothub.json]: Adding properties (even optional ones) to a resource types requires a new API version per the Azure Resource Providers versioning guidelines.

Comment (rkmanda) [inline arm-iothub/2016-02-03/swagger/iothub.json]: I have floated a mail internally to discuss the next steps for this

File: arm-iothub/2016-02-03/swagger/iothub.json
Comment (vishrutshah) [inline arm-iothub/2016-02-03/swagger/iothub.json]: If this is being set by server, `readOnly: true` should be removed. Similar to comment on line 1394.

File: arm-iothub/2016-02-03/swagger/iothub.json
Comment (vishrutshah) [inline arm-iothub/2016-02-03/swagger/iothub.json]: minor: description could be better phrased as:  "The name of the route. name can only include alphanumeric characters, periods, underscores, hyphens with maximum length of 64 characters and must be unique."

File: arm-iothub/2016-02-03/swagger/iothub.json
Comment (vishrutshah) [inline arm-iothub/2016-02-03/swagger/iothub.json]: Just curious: What are the reasons to model `endpointNames` as array of string that can only have 1 item?

**Review by vishrutshah — CHANGES_REQUESTED** [review]

File: arm-iothub/2016-02-03/swagger/iothub.json
Comment (rkmanda) [inline arm-iothub/2016-02-03/swagger/iothub.json]: Given that our RP is already live with this new property, we have a consensus now to do the following. 

- Rev the RP API version now, and once we do that, the new and old versions will start supporting the “routing” property.
- If a request comes for the old RP API version and the routing properties are set to null, we will merge the routing properties from the DB and act on the combined document. 
- If the request for the old RP API version comes with routing properties filled we will act on them as usual.

Going forward we will rev the API version for every new property introduced.

File: arm-iothub/2016-02-03/swagger/iothub.json
Comment (rkmanda) [inline arm-iothub/2016-02-03/swagger/iothub.json]: fixed the description based on the offline discussion we had

File: arm-iothub/2016-02-03/swagger/iothub.json
Comment (rkmanda) [inline arm-iothub/2016-02-03/swagger/iothub.json]: this allows for future expansion

File: arm-iothub/2016-02-03/swagger/iothub.json
Comment (rkmanda) [inline arm-iothub/2016-02-03/swagger/iothub.json]: removed

File: arm-iothub/2017-01-19/swagger/iothub.json
Comment (vishrutshah) [inline arm-iothub/2017-01-19/swagger/iothub.json]: what does minimun and maximum represent for "integer" type?

File: arm-iothub/2017-01-19/swagger/iothub.json
Comment (vishrutshah) [inline arm-iothub/2017-01-19/swagger/iothub.json]: why is this empty? Am i missing anything?

**Review by vishrutshah — CHANGES_REQUESTED** [review]

File: arm-iothub/2017-01-19/swagger/iothub.json
Comment (vishrutshah) [inline arm-iothub/2017-01-19/swagger/iothub.json]: What is the real purpose you are trying to achieve with defining 404 with schema?

Comment (rkmanda) [inline arm-iothub/2017-01-19/swagger/iothub.json]: The operationresults API in our service was implemented a while ago when the recommendation in the ARM spec was to place this API underneath the original request. 

https://<endpoint>/subscriptions/{subscriptionId}/{Original-URL-Fo-Request}/operationresults/{operationId}&api-version={api-version}

Due to this at the end of the long running delete ARM will throw a 404 because the original resource no longer exists. The new ARM recommendation is to move the operationresults API to 
underneath the subscription instead of the resource

https://<endpoint>/subscriptions/{subscriptionId}/ providers/{namespace}/locations/westus/operationresults/{operationId}&api-version={api-version}

However we haven't implemented this yet. Until then 404 will be a valid response for delete for us and we need this to be specified here to ensure good behavior for the autorest generated SDKs

File: arm-iothub/2017-01-19/swagger/iothub.json
Comment (rkmanda) [inline arm-iothub/2017-01-19/swagger/iothub.json]: sorry, can u elaborate as to what is empty?

File: arm-iothub/2017-01-19/swagger/iothub.json
Comment (rkmanda) [inline arm-iothub/2017-01-19/swagger/iothub.json]: It was a recommendation from @amarzavery to add minimum and maximum wherever there is a clear specific range of values that are possible for integer values.

File: arm-iothub/2017-01-19/swagger/iothub.json
Comment (rkmanda) [inline arm-iothub/2017-01-19/swagger/iothub.json]: we went thru this during the review process for the earlier version of this document and we have a sign-off. Are there any strong concerns against this?

File: arm-iothub/2017-01-19/swagger/iothub.json
Comment (rkmanda) [inline arm-iothub/2017-01-19/swagger/iothub.json]: We will rev the api version if it comes to that when we expand this set. But for now we will have to leave this the way it is since the RP already has these changes live.

File: arm-iothub/2017-01-19/swagger/iothub.json
Comment (rkmanda) [inline arm-iothub/2017-01-19/swagger/iothub.json]: its a valid set of skus of skus that this hub can transition to

File: arm-iothub/2017-01-19/swagger/iothub.json
Comment (rkmanda) [inline arm-iothub/2017-01-19/swagger/iothub.json]: nope

File: arm-iothub/2017-01-19/swagger/iothub.json
Comment (vishrutshah) [inline arm-iothub/2017-01-19/swagger/iothub.json]: No strong concerns but this will definitely provide better developer work-flow from SDK perspective. It's mostly a recommendation at this point but in future if you decide to apply `x-ms-client-flatten` then that'll be a breaking change in SDK. Considering you are revising version this time, I think you have opportunity here if there already some breaking changes going in with this swagger.

File: arm-iothub/2017-01-19/swagger/iothub.json
Comment (vishrutshah) [inline arm-iothub/2017-01-19/swagger/iothub.json]: Seems like the comments went to the wrong place. Sorry about that. 

I was referring to line number 1928 `"security": [ ]`.

File: arm-iothub/2017-01-19/swagger/iothub.json
Comment (rkmanda) [inline arm-iothub/2017-01-19/swagger/iothub.json]: done

File: arm-iothub/2017-01-19/swagger/iothub.json
Comment (vishrutshah) [inline arm-iothub/2017-01-19/swagger/iothub.json]: I see that makes sense that generated code won't throw an exception as 404 has schema for now. But I just wanted to point out that when you implement and remove 404 that'll potentially be a breaking change as the return type would change. I am just pointing it out so that you know.

File: arm-iothub/2017-01-19/swagger/iothub.json
Comment (vishrutshah) [inline arm-iothub/2017-01-19/swagger/iothub.json]: okay

File: arm-iothub/2017-01-19/swagger/iothub.json
Comment (vishrutshah) [inline arm-iothub/2017-01-19/swagger/iothub.json]: okay. sounds good then

File: arm-iothub/2017-01-19/swagger/iothub.json
Comment (vishrutshah) [inline arm-iothub/2017-01-19/swagger/iothub.json]: cool

File: arm-iothub/2017-01-19/swagger/iothub.json
Comment (vishrutshah) [inline arm-iothub/2017-01-19/swagger/iothub.json]: Could you please explain why is the definition for "security" empty here?

**Review by vishrutshah — COMMENTED** [review]
Thanks for addressing most of the comments. From the linter perspective things looks good. 

@amarzavery Could you please also validate this from data model perspective if we need to. Once that's done we can approve and merge it in.

**Comment by vishrutshah** [comment]
@rkmanda As per our offline talk on whether should you check-in the swagger with api-version that has not been released yet, We have decided that you should create a topic branch(`arm-iothub`) into https://github.com/Azure/azure-rest-api-specs and make PR into that for the unreleased version (Here 2017-01-19) and maintain it. Once you are ready with the service of that api-version, you can push it to the `master` branch. That way `master` branch of this repository stays shippable at any given point-in-time. 

As the changes LGTM, Could you please split this PR into 2. One for changes into older version and send PR to master and one for newer version and send PR into topic branch [Feel free to create topic branch named `arm-iothub`].

**Comment by salameer** [comment]
@rkmanda changes have been requested since the 13th of January and there has been no response from your side since then. Kindly note that if there are no updates on this thread by the 22nd of March it will be closed

**Comment by salameer** [comment]
Closing as requested by Roopesh:

> Roopesh Manda "Tue 4/4/2017 1:32 PM" Please go ahead and close it .. I will open another one..

**Comment by liydu** [comment]
@salameer could you enlighten me the other issue number for tracking this feature?

**Pull request closed without merging**
